Issue #88: use standard logging instead of print

diego-plan9 opened the pull request:
## Description

* add `QuantumProgram.enable_logs()` and `disable_logs()` as the single point of entry for setting/unsetting the verbosity of the output, replacing the different `verbose` and `silent` parameters.
* add a `qiskit._logging` module that contains two convenience functions for setting up the loggers (format, handlers) to be used by the SDK, defaulting to sensible values.
* replace `print()` with `logger.info()` (or `logger.debug()` on the mapper)

Not included in this PR:
* replacement of `verbose/silent` and `print` in `qiskit.tools` (might be well suited for a follow-up first contribution PR)
* fine grained decision about what current output should be `INFO/WARNING/DEBUG`
* allowing the user to override the logging format (might be better tackled along with a bigger scope PR that introduces some sort of user configuration mechanism)

## Motivation and Context

As described in #88, there are a number of drawbacks to using `print` statements, specially on a SDK. Since a number of functions can emit log information, it changes the way "verbosity" is requested or specified by the user: now the user should call:
``` python
qp = QuantumProgram()
... (operations where he is not interested in the output)
qp.enable_logs()
... (from that point onward, the operations will be print its output)
qp.disable_logs()
.... (optionally, hide the output from that point onward)
```

This is mostly because it is not recommended that a [library modifies the configuration of the loggers directly](https://docs.python.org/3/howto/logging.html#configuring-logging-for-a-library):
> (...) the configuration of handlers is the prerogative of the application developer who uses your library. The application developer knows their target audience and what handlers are most appropriate for their application: if you add handlers ‘under the hood’, you might well interfere with their ability to carry out unit tests and deliver logs which suit their requirements.

but at the same time trying to cater to users who want an easy way to "toggle the output" of specific snippets of their code and mostly just use `QuantumObject` directly.

## How Has This Been Tested?
`make test`. I would love to have some feedback and extra manual testing as it is very likely that the regular output is a bit more verbose than it was and there might be some formatting issues.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

ewinston commented on doc/dev_introduction.rst:
```diff
@@ -69,3 +69,54 @@ Several unroller backends and their outputs are summarized here:
+For convenience, :py:class:`QuantumProgram <qiskit.QuantumProgram>` provides two convenience
+methods (:py:func:`enable_logs() <qiskit.QuantumProgram.enable_logs>` and
+:py:func:`disable_logs() <qiskit.QuantumProgram.disable_logs>`) that modify the handlers
+and the level of the `qiskit` logger. Using these methods might interfer with the global
```
interfer -> interfere

ewinston commented:
Looks good to me!

diego-plan9 commented:
Updated the PR with changes spawned from the most recent discussion:
* reverted `get_execution_list()` in order to use `print` directly, as it is meant to be used only during interactive sessions.
* update the default logger configuration so `INFO` messages are printed without the extra timestamp, module, and level prepended string.

diego-plan9 commented:
As a follow-up to this PR, we should eventually:
* tackle the remaining `print` statements in the following files. Some of them might be false positives:

```
qiskit/qasm/_node/_creg.py
qiskit/qasm/_node/_id.py
qiskit/qasm/_node/_indexedid.py
qiskit/qasm/_node/_intnode.py
qiskit/qasm/_node/_node.py
qiskit/qasm/_node/_qreg.py
qiskit/qasm/_node/_real.py
qiskit/qasm/_qasmlexer.py
qiskit/qasm/_qasmparser.py
qiskit/qasm/_qasm.py
qiskit/tools/apps/optimization.py
qiskit/tools/qcvv/tomography.py
qiskit/tools/qi/pauli.py
qiskit/tools/visualization.py
```

* decide the level of the messages on a case-by-case basis. The output might be too verbose for some operations, and only useful for debugging or getting very fine-grained information.
* revise the places where messages are emitted because of an Exception or an error, and consider using `logger.exception` in those cases.
